src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1119,50 +1057,82 @@ void CASESession::SendErrorMsg(SigmaErrorType errorCode)
+        ReturnErrorOnFailure(RetrieveIPK(fabricId, ipkSpan));
+    }
 
-        VerifyOrReturnError(kTrustedRootIdSize == suppTlvReader.GetLength(), CHIP_ERROR_INVALID_TLV_ELEMENT);
-        VerifyOrReturnError(suppTlvReader.GetType() == TLV::kTLVType_ByteString, CHIP_ERROR_WRONG_TLV_TYPE);
-        ReturnErrorOnFailure(suppTlvReader.GetBytes(trustedRootId, kTrustedRootIdSize));
+    ReturnErrorOnFailure(hmac.HMAC_SHA256(mIPK, sizeof(mIPK), destinationMessage, sizeof(destinationMessage), destinationId.data(),
+                                          destinationId.size()));
 
-        if (mOpCredSet->IsTrustedRootIn(trustedRoot))
-        {
-            if (!mTrustedRootId.empty())
-            {
-                chip::Platform::MemoryFree(const_cast<uint8_t *>(mTrustedRootId.data()));
-                mTrustedRootId = CertificateKeyId();
-            }
-            mTrustedRootId = CertificateKeyId(reinterpret_cast<const uint8_t *>(chip::Platform::MemoryAlloc(kTrustedRootIdSize)));
-            VerifyOrReturnError(!mTrustedRootId.empty(), CHIP_ERROR_NO_MEMORY);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CASESession::FindDestinationIdCandidate(const ByteSpan & destinationId, const ByteSpan & initiatorRandom)
+{
+    uint8_t nCertificateSets = mOpCredSet->GetCertCount();
+
+    for (uint8_t i = 0; i < nCertificateSets; ++i)
+    {
+        uint8_t candidate[kSHA256_Hash_Length] = { 0 };
+        CertificateKeyId trustedRootId;
+        ChipCertificateData nodeOperationalCertificate;
+        NodeId nodeId;
+        FabricId fabricId;
+
+        trustedRootId = mOpCredSet->GetTrustedRootId(i);
+
+        ReturnErrorOnFailure(DecodeChipCert(mOpCredSet->GetDevOpCred(trustedRootId), mOpCredSet->GetDevOpCredLen(trustedRootId),
+                                            nodeOperationalCertificate));
+
+        ReturnErrorOnFailure(nodeOperationalCertificate.mSubjectDN.GetCertChipId(nodeId));
+        ReturnErrorOnFailure(nodeOperationalCertificate.mSubjectDN.GetCertFabricId(fabricId));
 
-            memcpy(const_cast<uint8_t *>(mTrustedRootId.data()), trustedRoot.data(), trustedRoot.size());
+        {
+            MutableByteSpan candidateSpan(candidate, sizeof(candidate));
+            ReturnErrorOnFailure(GenerateDestinationID(initiatorRandom, &trustedRootId, nodeId, fabricId, candidateSpan));
+        }
 
+        if (memcmp(destinationId.data(), candidate, destinationId.size()) == 0)
+        {
+            mTrustedRootId = trustedRootId;
             break;
         }
     }
+
     VerifyOrReturnError(!mTrustedRootId.empty(), CHIP_ERROR_CERT_NOT_TRUSTED);
 
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR CASESession::ConstructSaltSigmaR2(const ByteSpan & rand, const P256PublicKey & pubkey, const uint8_t * ipk,
-                                             size_t ipkLen, MutableByteSpan & salt)
+CHIP_ERROR CASESession::ConstructSaltSigmaR2(const ByteSpan & rand, const Crypto::P256PublicKey & pubkey, const ByteSpan & ipk,
```

Comment:
Not an applicable diff line, but there are a not of CHIP_ERROR_NO_MEMORY errors for `writer.Fit()` failing that should be CHIP_ERROR_BUFFER_TOO_SMALL